src/ZoneServer/World/Maps/Pathfinder.cs

```diff
@@ -0,0 +1,266 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Melia.Shared.World;
+using Melia.Shared.Game.Const;
+using Yggdrasil.Util;
+
+namespace Melia.Zone.World.Maps
+{
+	public class Pathfinder
+	{
+		private readonly int _maxNodeExpand = 500;
+		private readonly Dictionary<SizeType, float> _entitySizeRadius = new Dictionary<SizeType, float>
+		{
+			{ SizeType.None, 0 },
+			{ SizeType.Hidden, 0 },
+			{ SizeType.S, 12 },
+			{ SizeType.M, 15 },
+			{ SizeType.L, 20 },
+			{ SizeType.XL, 40 },
+			{ SizeType.XXL, 40 }
+		};
+
+		private Ground _ground;
+
+		/// <summary>
+		/// Loads the ground
+		/// </summary>
+		public void Load(Ground ground)
+		{
+			_ground = ground;
+		}
+
+		/// <summary>
+		/// Finds a path from start position to the goal position using
+		/// A* algorithm. Returns List of valid positions to goal.
+		/// Last element is always the closest position to goal.
+		/// Returns empty list if no path can be found.
+		/// </summary>
+		/// <param name="start"></param>
+		/// <param name="goal"></param>
+		/// <param name="entitySize"></param>
+		/// <returns></returns>
+		public List<Position> FindPath(Position start, Position goal, SizeType entitySize = SizeType.M)
```

Comment:
The entity size parameter should be sanity checked, as you're using it as the index for accessing the dictionary. You're expecting it to be `>= None <= XXL`, but we know those aren't all of them, and passing in an unexpected value would cause exceptions.